libraries/AP_HAL_Linux/Imx_IOMCU.hpp

```diff
@@ -0,0 +1,274 @@
+		{
+			close(fd);
+			fd = -1;
+			buf_used = 0;
+			poll_fd.fd = -1;
+		}
+	}
+	int init(const char *uart_name, unsigned int baud = 115200)
+	{
+		fd = open(uart_name, O_RDWR | O_NOCTTY);
+		if (fd < 0) {
+			printf("Message : %s\n", strerror(errno));
+			return fd;
+		}
+		if (set_uart_baudrate(fd, baud) <= 0)
+		{
+			printf("Message : %s\n", strerror(errno));
+			close(fd);
+			fd = -1;
+			return -1;
+		}
+		poll_fd.events = POLLIN;
+		poll_fd.fd = fd;
+		return fd;
+	}
+	bool set_freq(uint32_t chmask, uint32_t freq[12])
+	{
+		uint32_t packet[13] = { 0 };
+		char packet2[128];
+		packet[0] = chmask;
+		for (int i = 0; i < 12; i++)
+		{
+			packet[i + 1] = freq[i];
+		}
+
+		return uart_send_packet(fd, PACEET_RECV_SET_PWM_FREQ, packet, packet2, 13 * 4);
+	}
+	bool set_duty(uint32_t chmask, uint32_t duty[12])
+	{
+		uint32_t packet[13] = { 0 };
+		char packet2[128];
+		packet[0] = chmask;
+		for (int i = 0; i < 12; i++)
+		{
+			packet[i + 1] = duty[i];
+		}
+
+		return uart_send_packet(fd, PACKET_RECV_SET_PWM_DUTY, packet, packet2, 13 * 4);
+	}
+	bool read_adc(int32_t *adc_pc0_data, int32_t *adc_pc1_data)
+	{
+		int poll_ret = poll(&poll_fd, 1, 0);
+		if (poll_ret <= 0)
+			return false;
+
+		int read_len = read(fd, &uart_buf[buf_used], 4096 - buf_used);
+		if (read_len <= 0)
+		{
+			return false;
+		}
+		buf_used += read_len;
+
+		uint8_t packet_id;
+		uint8_t packet[4096];
+		uint8_t packet_size;
+		if (recv_packet(uart_buf, &buf_used, &packet_id, packet, &packet_size) == false)
+		{
+			return false;
+		}
+		if (PACKET_SEND_ADC_VALUE != packet_id)
+			return false;
+
+		adc_value_s *padc_value = (adc_value_s *)packet;
+		*adc_pc0_data = padc_value->adc_pc0_data;
+		*adc_pc1_data = padc_value->adc_pc1_data;
+
+		return true;
+	}
+private:
+	uint8_t uart_buf[4096];
+	int buf_used;
+	struct pollfd poll_fd;
+
+	int set_uart_baudrate(const int _fd, unsigned int baud)
+	{
+		int speed;
+		switch (baud) {
+		case 9600:   speed = B9600;   break;
+		case 19200:  speed = B19200;  break;
+		case 38400:  speed = B38400;  break;
+		case 57600:  speed = B57600;  break;
+		case 115200: speed = B115200; break;
+		case 230400: speed = B230400; break;
+		default:
+			return -EINVAL;
+		}
+
+		struct termios uart_config;
+
+		int termios_state;
+
+		tcgetattr(_fd, &uart_config);
+
+		uart_config.c_cflag |= (CLOCAL | CREAD);
+		uart_config.c_cflag &= ~PARENB;
+		uart_config.c_cflag &= ~CSTOPB;
+		uart_config.c_cflag &= ~CSIZE;
+		uart_config.c_cflag |= CS8;
+		uart_config.c_cflag &= ~CRTSCTS;
+
+		uart_config.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
+
+		uart_config.c_iflag = 0;
+
+		uart_config.c_oflag = 0;
+
+		uart_config.c_cc[VTIME] = 0;
+		uart_config.c_cc[VMIN] = 1;
+
+		if ((termios_state = cfsetispeed(&uart_config, speed)) < 0) {
+			return 0;
+		}
+
+		if ((termios_state = cfsetospeed(&uart_config, speed)) < 0) {
+			return 0;
+		}
+
+		if ((termios_state = tcsetattr(_fd, TCSANOW, &uart_config)) < 0) {
+			return 0;
+		}
+
+		return 1;
```

Comment:
Take a look on UARTDevice. We have all of these implemented there.

Reply:
i have read UARTDevice,But I don't think I need such a complex class.
Do I have to change the code?